### ISSUE TITLE
Type checker: do not use ``validForCalldata`` for memory arrays or rename it.

### ISSUE THREAD

**Issue by ekpyron** [ethereum/solidity#7498]
Came up in https://github.com/ethereum/solidity/pull/7496.
The fix in https://github.com/ethereum/solidity/pull/7496 allows to use ``isDynamicallyEncoded`` for types without interface types (i.e. recursive structs), which is somewhat counter-intuitive.
The reason why that's necessary is that the type checker uses ``validForCalldata`` for memory arrays. We should investigate which part of ``validForCalldata`` applies to memory arrays and adjust the check accordingly.
EDIT: the main reason to call the function is to prevent extremely large memory arrays - if we want to exclude them we should do that by ways of ``memoryStride()``, etc., instead of reusing the calldata code, though.

**Comment by ekpyron** [comment]
Potential simple solution: rename ``validForCalldata`` to something like ``hasSufficientlySmallStaticSize`` (name to be improved upon ;-)) and use ``unlimitedStaticSize`` (to be implemented) which uses ``memoryStride`` instead of ``calldataStride`` for memory arrays.

**Comment by ekpyron** [comment]
Once this is done: investigate whether we can remove ``if (recursive()) return true;`` from ``StructType::isDynamicallyEncoded`` again.

**Comment by chriseth** [comment]
I don't think we can, since the fact that some part anywhere inside the directed graph has a too large static size is already a problem. We particularly do not want that in storage. Thinking of it: The fix we implemented might not have fixed the underlying problem...

**Comment by ekpyron** [comment]
> We particularly do not want that in storage.

The particular check, i.e. ``validForCalldata()`` is only used for memory and calldata, not for storage it seems (TypeChecker.cpp:L502-507)... not sure what and where we do this for storage.

> Thinking of it: The fix we implemented might not have fixed the underlying problem...

That's true and why I created this issue.
But I think we should be fine, if we reimplement ``validForCalldata`` as ``validForMemory`` (or with one shared implementation) and just use the ``memory*Size`` functions instead of the calldata ones for the static size calculation...

**Comment by ekpyron** [comment]
The underlying issue of the particular issue as far as I understand it is just that we use the calldata size functions to enforce an upper limit on memory size and the calldata size functions asserted interface types, which are not present in some types which are still viable for memory, in particular recursive structs...

There may be a separate issue that checking for size bounds "locally" once in every "level" of the array is insufficient and that we have to consider full paths instead, but I'm not sure - I think it might actually be fine, since the size functions should be correct for every statically sized type and for dynamically encoded ones only the pointer size is relevant anyways...

**Comment by ekpyron** [comment]
Done in https://github.com/ethereum/solidity/pull/8665
